schema/validator.go

```diff
@@ -73,3 +86,35 @@ type unimplemented string
 func (v unimplemented) Validate(src io.Reader) error {
 	return fmt.Errorf("%s: unimplemented", v)
 }
+
+func validateManifestDescendants(r io.Reader) error {
+	header := v1.Manifest{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
+	if err != nil {
+		return errors.Wrap(err, "manifest format mismatch")
+	}
+
+	if header.Config.MediaType != string(v1.MediaTypeImageConfig) &&
+		header.Config.MediaType != MediaTypeDockerImageConfig {
+		return fmt.Errorf("illegal config mediaType: %s", header.Config.MediaType)
+	}
+
+	for _, layer := range header.Layers {
+		if layer.MediaType != string(v1.MediaTypeImageLayer) &&
+			layer.MediaType != MediaTypeDockerImageLayer &&
+			layer.MediaType != MediaTypeNonDistributable {
+			return fmt.Errorf("illegal layer mediaType: %s", layer.MediaType)
```

Comment:
On Wed, Oct 05, 2016 at 07:10:01PM -0700, xiekeyang wrote:
> 1. Do you think we should emit warn for all string not equal to OCI type, even for buzz words? It seems too loose.

We're warning them, and they would be wise to take the warning seriously.  I don't think we can do anything stronger by default; how do we know that the strange type they're using *isn't* handled by their tooling?  But I think image-tools should have a `--strict` flag (for “error on any features not required in compliant unpackers”) and/or an [`--unpackable` flag][1] (for “error on any features *I* don't recognize”).

[1]: https://github.com/opencontainers/image-spec/issues/286#issuecomment-245856987

> 2. Current logger have no `warn` level. Do you agree that we import `logrus` here, to print different  levels log?

I think [logging is orthogonal][1].  These tests are the primary output of validation, so I think a validator should be printing the validation results (including warnings like this) to stdout (or some sort of writer).  In my callback-based mock-up, I'm [using `tap.go`][2], and in that case warnings would [look like][3]:

    ok 7 # skip sha256-a… layers[2] has an unknown media type: image/png

[1]: https://github.com/opencontainers/image-spec/pull/234#issuecomment-245949723
[2]: https://github.com/wking/image-tools/commit/4f0b3c1bae97d212680dc73ee8b35b6e47b8a8c6#diff-201d416d61af62a246e5b552b6d42c92R31
[3]: https://github.com/opencontainers/image-spec/issues/286#issuecomment-245876396